packages/@aws-cdk/aws-applicationautoscaling/lib/scalable-target.ts

```diff
@@ -0,0 +1,211 @@
+
+/**
+ * Properties for a scalable target
+ */
+export interface ScalableTargetProps {
+  /**
+   * The minimum value that Application Auto Scaling can use to scale a target during a scaling activity.
+   */
+  minCapacity: number;
+
+  /**
+   * The maximum value that Application Auto Scaling can use to scale a target during a scaling activity.
+   */
+  maxCapacity: number;
+
+  /**
+   * Role that allows Application Auto Scaling to modify your scalable target.
+   *
+   * @default A role is automatically created
+   */
+  role?: iam.IRole;
+
+  /**
+   * The resource identifier to associate with this scalable target.
+   *
+   * This string consists of the resource type and unique identifier.
+   *
+   * @example service/ecsStack-MyECSCluster-AB12CDE3F4GH/ecsStack-MyECSService-AB12CDE3F4GH
+   * @see https://docs.aws.amazon.com/autoscaling/application/APIReference/API_RegisterScalableTarget.html
+   */
+  resourceId: string;
+
+  /**
+   * The scalable dimension that's associated with the scalable target.
+   *
+   * Specify the service namespace, resource type, and scaling property.
+   *
+   * @example ecs:service:DesiredCount
+   * @see https://docs.aws.amazon.com/autoscaling/application/APIReference/API_ScalingPolicy.html
+   */
+  scalableDimension: string;
+
+  /**
+   * The namespace of the AWS service that provides the resource or
+   * custom-resource for a resource provided by your own application or
+   * service.
+   *
+   * For valid AWS service namespace values, see the RegisterScalableTarget
+   * action in the Application Auto Scaling API Reference.
+   *
+   * @see https://docs.aws.amazon.com/autoscaling/application/APIReference/API_RegisterScalableTarget.html
+   */
+  serviceNamespace: string;
+}
+
+/**
+ * Define a scalable target
+ */
+export class ScalableTarget extends cdk.Construct {
+  /**
+   * ID of the Scalable Target
+   *
+   * @example service/ecsStack-MyECSCluster-AB12CDE3F4GH/ecsStack-MyECSService-AB12CDE3F4GH|ecs:service:DesiredCount|ecs
+   */
+  public readonly scalableTargetId: string;
+
+  /**
+   * The role used to give AutoScaling permissions to your resource
+   */
+  public readonly role: iam.IRole;
+
+  private readonly actions = new Array<cloudformation.ScalableTargetResource.ScheduledActionProperty>();
+
+  constructor(parent: cdk.Construct, id: string, props: ScalableTargetProps) {
+    super(parent, id);
+
+    if (props.maxCapacity < 0) {
+      throw new RangeError(`maxCapacity cannot be negative, got: ${props.maxCapacity}`);
+    }
+    if (props.minCapacity < 0) {
+      throw new RangeError(`minCapacity cannot be negative, got: ${props.minCapacity}`);
+    }
+    if (props.maxCapacity < props.minCapacity) {
+      throw new RangeError(`minCapacity (${props.minCapacity}) should be lower than maxCapacity (${props.maxCapacity})`);
+    }
+
+    this.role = props.role || new iam.Role(this, 'Role', {
```

Comment:
Can we allow service-linked role only?

Application Auto Scaling provides service-linked roles for targets the service supports: https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-service-linked-roles.html

Reply:
What's the harm in allowing a Role to be used? According to the docs, not all services work properly with a Service Linked Role, and this mechanism is consistent with other places we require roles in the CDK.

Reply:
Please correct me if I am wrong in the below comments.

> What's the harm in allowing a Role to be used?

I am not sure if allowing any Role or creating a new Role is scalable. What if there are 2,000 scalable targets, which is practical use case in ECS auto scaling? I would assume that users can create 2,000 roles and pass them if they misuse it.

Additionally, if any Role is allowed, it would mean any policy can attach to the role. If users miss adding some permissions to the policy, then auto scaling won't work. In case of service-linked role, a proper policy managed by AWS is already attached so that we can guarantee auto scaling functionality in terms of permissions.

>  According to the docs, not all services work properly with a Service Linked Role

Could you share a link for the docs if possible? Does "all services" means "all targets of Application Auto Scaling" or "all AWS services"? I think I don't get this part.

> this mechanism is consistent with other places we require roles in the CDK.

In this sense, I don't mind allowing Role. If possible, however, it would be great to have comment or documentation saying "please consider service-linked role as well".

Reply:
Just going off what it says here:

https://docs.aws.amazon.com/autoscaling/application/APIReference/API_RegisterScalableTarget.html#autoscaling-RegisterScalableTarget-request-RoleARN

> For resources that are not supported using a service-linked role, this parameter is required and must specify the ARN of an IAM role

Reply:
I see. Thanks.